lib/nostrum/shard/session.ex

```diff
@@ -86,63 +111,63 @@ defmodule Nostrum.Shard.Session do
 
     case from_handle do
       {new_state, reply} ->
-        {:reply, {:binary, reply}, new_state}
+        :ok = :gun.ws_send(state.conn, {:binary, reply})
+        {:noreply, new_state}
 
       new_state ->
-        {:ok, new_state}
+        {:noreply, new_state}
     end
   end
 
+  def handle_info({:gun_ws, _conn, _stream, {:close, errno, reason}}, state) do
+    Logger.info("Shard websocket closed (errno #{errno}, reason #{inspect(reason)}).")
+    {:noreply, state}
+  end
+
+  def handle_info(
+        {:gun_down, _conn, _proto, _reason, _killed_streams, _unprocessed_streams},
+        state
+      ) do
+    {:noreply, state}
+  end
+
+  def handle_info({:gun_up, worker, _proto}, state) do
+    stream = :gun.ws_upgrade(worker, @gateway_qs)
+    await_ws_upgraded(worker, stream)
+    Logger.warn("Reconnected after connection broke.")
+    {:noreply, state}
+  end
+
   def handle_cast({:status_update, payload}, state) do
-    {:reply, {:binary, payload}, state}
+    :ok = :gun.ws_send(state.conn, {:binary, payload})
+    {:noreply, state}
   end
 
   def handle_cast({:update_voice_state, payload}, state) do
     {:reply, {:binary, payload}, state}
   end
 
   def handle_cast({:request_guild_members, payload}, state) do
-    {:reply, {:binary, payload}, state}
+    :ok = :gun.ws_send(state.conn, {:binary, payload})
+    {:noreply, state}
   end
 
   def handle_cast(:heartbeat, %{heartbeat_ack: false} = state) do
     Logger.warn("heartbeat_ack not received in time, disconnecting")
-    {:close, state}
+    :ok = :gun.close(state.conn)
+    {:stop, {:heartbeat_ack_timeout, state.shard_num}, state}
   end
 
   def handle_cast(:heartbeat, state) do
```

Comment:
We need to keep in mind how this will work. Right now if we successfully resume we will have two heartbeats because we don't cancel the timer on disconnect and for every hello we start heartbeating again. 

We should either:
 * Before we start a new timer, check if one currently exists, and if so use that one. This is potentially problematic if things line up weirdly as I'm not sure if discord expects the first heartbeat immediately or after the interval upon connection.

* Cancel the timer on disconnect and our logic should restart it when we get another hello message from discord.

Reply:
We don't resume successfully right now though...

Reply:
Because of the `:data_error` issue? While that may be the case I think your current implementation is still wrong.